### DISCUSSION TITLE
Type inference for forward references?

### DISCUSSION THREAD

**Discussion by vremip** [microsoft/pylance-release#3274] [Q&A]
Referring to [PEP 484](https://peps.python.org/pep-0484/#forward-references), on the subject of forward references, one can use "module.class" for a valid type hinting. Pylance interprets the type correctly.

```
# File models/a.py
from models import b
class A(Model):
    def foo(self, argument: 'b.B'): ...

# File models/b.py
from models import a
class B(Model):
    def bar(self, argument: 'a.A'): ...
```

For various reasons, in our project, we'd rather want the imports to be more like this: 

```
# File models/a.py
import models
class A(Model):
    def foo(self, argument: 'models.b.B'): ...

# File models/b.py
import models
class B(Model):
    def bar(self, argument: 'models.a.A'): ...
```

If I do this, the code compiles without a hitch, but **Pylance** will not infer the types in *vs-code*, meaning "argument.bar" in the method "A.foo" will be grayed out, making development more difficult.

Is there a setting to enable this type of hinting? Thank you

**Comment by erictraut** [comment]
Both of these approaches should work fine. We'll need additional details to diagnose the problem you're seeing. In your example, you are importing from a module called `models`, but you don't indicate where that comes from. Is it part of an installed library? If so, is it part of a "py.typed" library? 

In your examples above, none of these annotations are forward references, so using quotes for the type annotations is unnecessary. Is there a reason you want to use quotes here?

You might also want to consider the use of `from __future__ import annotations` to avoid the need to quote forward references in annotations.
